opendmarc/opendmarc-arcseal.c

```diff
@@ -0,0 +1,267 @@
+	{ "i",		AS_TAG_INSTANCE },
+    { "d",		AS_TAG_SIGNATURE_DOMAIN },
+	{ "s",		AS_TAG_SIGNATURE_SELECTOR },
+	{ "t",		AS_TAG_SIGNATURE_TIME },
+    { "b",		AS_TAG_SIGNATURE_VALUE },
+	{ NULL,		AS_TAG_UNKNOWN }
+};
+
+/*
+**  ARC_SEAL_CONVERT -- convert a string to its code
+**
+**  Parameters:
+**  	table -- in which table to look up
+**  	str -- string to find
+**
+**  Return value:
+**  	A code translation of "str".
+*/
+
+static int
+arc_seal_convert(struct arc_seal_lookup *table, char *str)
+{
+	int c;
+
+	assert(table != NULL);
+	assert(str != NULL);
+
+	for (c = 0; ; c++)
+	{
+		if (table[c].str == NULL || strcasecmp(table[c].str, str) == 0)
+        	return table[c].code;
+	}
+
+	/* NOTREACHED */
+}
+
+
+/*
+**  ARC_SEAL_STRIP_WHITESPACE -- removes all whitespace from a string in-place,
+**	                  handling a maximum string of length ARC_SEAL_MAX_TOKEN_LEN
+**
+**  Parameters:
+**  	string -- NULL-terminated string to modify
+**
+**  Returns:
+**  	pointer to string on success, NULL on failure (max string length
+**		              exceeded)
+**/
+
+static char *
+arc_seal_strip_whitespace(u_char *string)
+{
+	assert(string != NULL);
+
+	char *string_ptr = string;
+	char space_chars[] = { ' ', '\t', '\r', '\n' };
+	int space_chars_count = 4;
+	int a, b, c;
+	_Bool space_found;
+
+	for (a = 0, b = 0; string[b] != '\0' && b < ARC_SEAL_MAX_TOKEN_LEN; b++)
+	{
+		space_found = FALSE;
+		for(c = 0; c < sizeof space_chars; c++)
+		{
+			if (string[b] == space_chars[c])
+			{
+				space_found = TRUE;
+				break;
+			}
+		}
+
+		if (space_found)
+		{
+			continue;
+		}
+
+		string[a] = string[b];
+		a++;
+	}
+
+	if (b >= ARC_SEAL_MAX_TOKEN_LEN)
+	{
+		return NULL;
+	}
+
+	/* set remaining chars to null */
+	memset(&string[a], '\0', sizeof(char) * (b - a));
+
+	return string;
+}
+
+/*
+**  ARC_SEAL_STRIP_FIELD_NAME -- strip the field name from a string, skipping
+**	                  leading whitespace
+**
+**  Parameters:
+**  	field -- NULL-terminated string
+**		name -- NULL-terminated string containing field to remove
+**		delim -- NULL-terminated string containing delimiter
+**  	buf -- destination buffer
+**  	buflen -- number of bytes at buf
+**
+**  Returns:
+**  	0 on success, -1 on failure
+**/
+
+static int
+arc_seal_strip_field_name(u_char *field, u_char *name, u_char *delim, char *buf, size_t buf_len)
+{
+    size_t copy_len;
+
+    assert(field != NULL);
+    assert(name != NULL);
+	assert(delim != NULL);
+    assert(buf != NULL);
+    assert(buf_len > 0);
+
+    size_t name_len = strlen(name);
+    size_t delim_len = strlen(delim);
+
+    if (name_len + delim_len > ARC_SEAL_MAX_FIELD_NAME_LEN)
+    {
+        return -1;
+    }
+
+    /* build delimited name */
+    u_char name_delim[ARC_SEAL_MAX_FIELD_NAME_LEN + 1];
+    memcpy(name_delim, (const void *)name, name_len);
+    memcpy(name_delim + name_len, delim, delim_len);
+    memset(name_delim + name_len + delim_len, '\0', sizeof(char));
```

Comment:
name_delim could be omitted; all you need is name_len + delim_len on the next line.